File: scripts/safe_browser.py
import asyncio
import sys
import argparse
import nodriver as uc
import time
import json

async def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--token', default='', help='auth_token or json cookie dict (raw)')
    parser.add_argument('--token_b64', default='', help='base64 encoded token')
    parser.add_argument('--platform', default='twitter', help='twitter or facebook')
    args = parser.parse_args()

    import base64
    if args.token_b64:
        auth_token = base64.b64decode(args.token_b64).decode('utf-8')
    else:
        auth_token = args.token

    platform = args.platform

    browser = None
    try:
        # Start browser in interactive mode (headless=False) so user can browse
        browser = await uc.start(headless=False)
        
        if platform == 'facebook':
            page = await browser.get('https://facebook.com')
            try:
                c_user = ""
                xs = ""
                try:
                    cookies = json.loads(auth_token)
                    c_user = str(cookies.get('c_user', ''))
                    xs = str(cookies.get('xs', ''))
                except json.JSONDecodeError:
                    import re
                    c_match = re.search(r'c_user=([^;]+)', auth_token)
                    x_match = re.search(r'xs=([^;]+)', auth_token)
                    if c_match: c_user = c_match.group(1)
                    if x_match: xs = x_match.group(1)
                    
                if c_user:
                    await page.send(uc.cdp.network.set_cookie(
                        name='c_user', value=c_user, domain='.facebook.com', path='/',
                        secure=True, http_only=False
                    ))
                if xs:
                    await page.send(uc.cdp.network.set_cookie(
                        name='xs', value=xs, domain='.facebook.com', path='/',
                        secure=True, http_only=True
                    ))
            except Exception as e:
                print(f"Failed to parse facebook cookies: {str(e)}")
            
            await page.get('https://facebook.com/')
            
        else: # Default Twitter
            page = await browser.get('https://x.com')
            
            # Create cookie object and set it
            await page.send(uc.cdp.network.set_cookie(
                name='auth_token',
                value=auth_token,
                domain='.x.com',
                path='/',
                secure=True,
                http_only=True
            ))

            # Navigate direct to home feed with the newly injected auth token
            await page.get('https://x.com/home')
            
        print("Browser session started successfully. Keeping alive...")
        sys.stdout.flush()

        # Keep script running infinitely so the browser window stays open for the user.
        # If user manually closes the window, the script will naturally terminate eventually.
        while True:
            await asyncio.sleep(60)

    except Exception as e:
        print(f"Error launching safe browser: {str(e)}")
        sys.exit(1)
    finally:
        # Do not automatically close the browser here, unless it crashed.
        # The user's whole goal is to use the browser manually.
        pass

if __name__ == '__main__':
    asyncio.run(main())

File: scripts/safe_browser.js
import puppeteer from 'puppeteer-extra';
import StealthPlugin from 'puppeteer-extra-plugin-stealth';

puppeteer.use(StealthPlugin());

async function main() {
    const args = process.argv.slice(2);
    
    // Parse arguments
    let b64Token = '';
    let platform = 'twitter';
    
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--token_b64' && args[i+1]) {
            b64Token = args[i+1];
        } else if (args[i] === '--platform' && args[i+1]) {
            platform = args[i+1];
        }
    }

    if (!b64Token) {
        console.error("No token provided");
        process.exit(1);
    }

    const auth_token = Buffer.from(b64Token, 'base64').toString('utf8');

    const browser = await puppeteer.launch({
        headless: false,
        defaultViewport: null,
        args: [
            '--start-maximized',
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-blink-features=AutomationControlled',
            '--disable-notifications'
        ]
    });

    const pages = await browser.pages();
    const page = pages.length > 0 ? pages[0] : await browser.newPage();

    if (platform === 'facebook') {
        let parsedCookies = [];
        try {
            const cookiesData = JSON.parse(auth_token);
            if (Array.isArray(cookiesData)) {
                parsedCookies = cookiesData;
            } else {
                // Legacy object format if applicable
                if (cookiesData.c_user && cookiesData.xs) {
                    parsedCookies = [
                        { name: 'c_user', value: String(cookiesData.c_user), domain: '.facebook.com', path: '/', secure: true, httpOnly: false, sameSite: 'None' },
                        { name: 'xs', value: String(cookiesData.xs), domain: '.facebook.com', path: '/', secure: true, httpOnly: true, sameSite: 'None' }
                    ];
                }
            }
        } catch (e) {
            let c_user = '';
            let xs = '';
            const cMatch = auth_token.match(/c_user=([^;\s]+)/);
            const xMatch = auth_token.match(/xs=([^;\s]+)/);
            if (cMatch) c_user = cMatch[1];
            if (xMatch) xs = xMatch[1];
            if (c_user && xs) {
                parsedCookies = [
                    { name: 'c_user', value: String(c_user), domain: '.facebook.com', path: '/', secure: true, httpOnly: false, sameSite: 'None' },
                    { name: 'xs', value: String(xs), domain: '.facebook.com', path: '/', secure: true, httpOnly: true, sameSite: 'None' }
                ];
            }
        }

        if (parsedCookies.length > 0) {
            await page.setCookie(...parsedCookies);
        }
        await page.goto('https://www.facebook.com/');
    } else {
        await page.setCookie({
            name: 'auth_token',
            value: String(auth_token),
            domain: '.twitter.com',
            path: '/',
            secure: true,
            httpOnly: true,
            sameSite: 'None'
        });
        await page.setCookie({
            name: 'auth_token',
            value: String(auth_token),
            domain: '.x.com',
            path: '/',
            secure: true,
            httpOnly: true,
            sameSite: 'None'
        });
        await page.goto('https://x.com/home');
    }
    
    console.log("Browser launched successfully. Keeping alive...");
    
    browser.on('disconnected', () => {
        console.log("Browser closed. Exiting.");
        process.exit(0);
    });
}

main().catch(error => {
    console.error("Failed to start browser:", error);
    process.exit(1);
});
